### PR TITLE
[CC][GFC-82]Added in ValidIf to frontend

### PR DESCRIPTION
**Pull request by cavster**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/validation/ValidationService.scala [hunk 1]
```diff
@@ -49,12 +50,12 @@ class ValidationService(
     repeatService: RepeatingComponentService
 ) {
 
-  private def validateFieldValue(fieldValue: FormComponent, data: Map[FormComponentId, Seq[String]], envelopeId: EnvelopeId)(implicit hc: HeaderCarrier): Future[ValidatedType] =
-    new ComponentsValidator(data, fileUploadService, envelopeId).validate(fieldValue)
+  private def validateFieldValue(fieldValue: FormComponent, data: Map[FormComponentId, Seq[String]], envelopeId: EnvelopeId, retrivals: Retrievals)(implicit hc: HeaderCarrier): Future[ValidatedType] =
+    new ComponentsValidator(data, fileUploadService, envelopeId, retrivals).validate(fieldValue)
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/validation/ValidationService.scala]: Please fix the misspelling of retrivals

File: app/uk/gov/hmrc/gform/validation/ValidationService.scala [hunk 2]
```diff
@@ -49,12 +50,12 @@ class ValidationService(
...
 
-  def validateComponents(fieldValues: List[FormComponent], data: Map[FormComponentId, Seq[String]], envelopeId: EnvelopeId)(implicit hc: HeaderCarrier): Future[ValidatedType] =
+  def validateComponents(fieldValues: List[FormComponent], data: Map[FormComponentId, Seq[String]], envelopeId: EnvelopeId, retrivals: Retrievals)(implicit hc: HeaderCarrier): Future[ValidatedType] =
     fieldValues
-      .map(fv => validateFieldValue(fv, data, envelopeId))
+      .map(fv => validateFieldValue(fv, data, envelopeId, retrivals))
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/validation/ValidationService.scala]: Ditto

File: app/uk/gov/hmrc/gform/validation/ValidationService.scala [hunk 3]
```diff
@@ -234,27 +236,28 @@ class ComponentsValidator(data: Map[FormComponentId, Seq[String]], fileUploadSer
...
-      case (_, value :: Nil, TelephoneNumber) => textValidator(fieldValue, value, ValidationValues.phoneDigits._1, ValidationValues.phoneDigits._2)
-      case (_, value :: Nil, Email) => Monoid.combine(email(fieldValue, value), textValidator(fieldValue, value, 0, ValidationValues.emailLimit))
-      case (_, value :: Nil, Number(maxWhole, maxFractional, _)) => validateNumber(fieldValue, value, maxWhole, maxFractional, false)
-      case (_, value :: Nil, PositiveNumber(maxWhole, maxFractional, _)) => validateNumber(fieldValue, value, maxWhole, maxFractional, true)
-      case (false, Nil, _) => ().valid
-      case (_, value :: rest, _) => ().valid // we don't support multiple values yet
+    (fieldValue.mandatory, fieldValue.validIf, textData.filterNot(_.isEmpty()), text.constraint) match {
+      case (true, _, Nil, _) => getError(fieldValue, "Please enter required data")
+      case (_, _, _, AnyText) => ().valid
+      case (_, Some(validIfExpression), _, _) => checkValidIf(fieldValue, data, validIfExpression.expr)
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/validation/ValidationService.scala]: Ah, no, this doesn't do what is needed.

If there is a validIf and a text constraints, then we need to check both the validIf **and** the constraint

Comment (cavster) [reply app/uk/gov/hmrc/gform/validation/ValidationService.scala]: That's true as long as they set what it has to be valid we should fine.Better to be on the same side I suppose ¯\_(ツ)_/¯

**Pull request merged**